collie/model/hybrid_matrix_factorization.py

```diff
@@ -264,29 +362,62 @@ def _setup_model(self, **kwargs) -> None:
                                                embedding_dim=self.hparams.embedding_dim)
         self.dropout = nn.Dropout(p=self.hparams.dropout_p)
 
-        # set up metadata-only layers
-        metadata_output_dim = self.hparams.item_metadata_num_cols
-        self.metadata_layers = None
-        if self.hparams.metadata_layers_dims is not None:
-            metadata_layers_dims = (
-                [self.hparams.item_metadata_num_cols] + self.hparams.metadata_layers_dims
+        # set up item metadata-only layers
+        item_metadata_output_dim = self.hparams.item_metadata_num_cols
+        self.item_metadata_layers = None
+
+        if self.hparams.item_metadata_layers_dims is not None:
+            item_metadata_layers_dims = (
+                [self.hparams.item_metadata_num_cols] + self.hparams.item_metadata_layers_dims
             )
-            self.metadata_layers = [
-                nn.Linear(metadata_layers_dims[idx - 1], metadata_layers_dims[idx])
-                for idx in range(1, len(metadata_layers_dims))
+            self.item_metadata_layers = [
+                nn.Linear(item_metadata_layers_dims[idx - 1], item_metadata_layers_dims[idx])
+                for idx in range(1, len(item_metadata_layers_dims))
             ]
-            for i, layer in enumerate(self.metadata_layers):
-                nn.init.xavier_normal_(self.metadata_layers[i].weight)
-                self.add_module('metadata_layer_{}'.format(i), layer)
-
-            metadata_output_dim = metadata_layers_dims[-1]
-
-        # set up combined layers
-        combined_dimension_input = (
-            self.user_embeddings.embedding_dim
-            + self.item_embeddings.embedding_dim
-            + metadata_output_dim
-        )
+            for i, layer in enumerate(self.item_metadata_layers):
+                nn.init.xavier_normal_(self.item_metadata_layers[i].weight)
+                self.add_module('item_metadata_layer_{}'.format(i), layer)
+
+            item_metadata_output_dim = item_metadata_layers_dims[-1]
```

Comment:
[Optional] This seems like a good chunk of code that we have to repeat for the user side with not a ton of differences. I wonder if it's worth splitting into a helper method that we can just call once for item and once for user metadata architecture construction.

Reply:
@nathancooperjones Tried something here and with the other optional helper function in the `forward` function and I am not quite sure how to best proceed. My main struggle is the issue with `item` and `user` specific variables that get created that we want to move into a helper function. One idea I had is to create a dictionary to make it easier to deal with these `item` and `user` specific variables, but I wanted to hear your ideas before I go further.

Reply:
What about something like this: 

```python
def _configure_metadata_layers(
    self,
    metadata_layers_attr: str,
    metadata_layer_name: str,
    metadata_layers_dims: Optional[Iterable[int]],
    num_metadata_cols: Optional[int],
) -> None:
    """
    Configure metadata layers for either item or user data.

    Parameters
    ----------
    metadata_layers_attr: str
        Attribute name for the metadata layers (e.g. ``item_metadata_layers``)
    metadata_layer_name: str
        Name for the added module used in model construction (e.g. ``item_metadata_layer``). Note
        that in the actual model, the name will be appended with an ``_`` and a numbered index,
        starting at ``0``
    metadata_layers_dims: list
        List of dimensions for the hidden state of the metadata layers
    num_metadata_cols: int
        Number of columns in the metadata dataset

    """
    if metadata_layers_dims is not None:
        full_metadata_layers_dims = (
            [num_metadata_cols] + metadata_layers_dims
        )

        full_metadata_layers = [
            nn.Linear(full_metadata_layers_dims[idx - 1], full_metadata_layers_dims[idx])
            for idx in range(1, len(full_metadata_layers_dims))
        ]

        setattr(self, 'metadata_layers_attr', full_metadata_layers)

        for i, layer in enumerate(getattr(self, metadata_layers_attr)):
            nn.init.xavier_normal_(getattr(self, metadata_layers_attr)[i].weight)
            self.add_module('{metadata_layer_name}_{}'.format(i), layer)


self._configure_metadata_layers(
    metadata_layers_attr='item_metadata_layers',
    metadata_layer_name='item_metadata_layer',
    metadata_layers_dims=self.hparams.item_metadata_layers_dims,
    num_metadata_cols=self.hparams.item_metadata_num_cols,
)

self._configure_metadata_layers(
    metadata_layers_attr='user_metadata_layers',
    metadata_layer_name='user_metadata_layer',
    metadata_layers_dims=self.hparams.user_metadata_layers_dims,
    num_metadata_cols=self.hparams.user_metadata_num_cols,
)
```